### ISSUE TITLE
Implement Mouse Relative Mode in NoGUI for Mouse Playstation AND for GunCon in both QT and NoGUI

### ISSUE THREAD

**Issue by LukasThyWalls** [stenzek/duckstation#2104]
**DESCRIPTION**:

I did some test with:
* Resident Evil Survivor (EU/PAL) SLES-02732
* Command & Conquer: Red Alert - Disc 2 [Soviet] (USA/NTSC) SLUS-00485

Both can be played with the mouse with two different controllers: C&C with the Playstation Mouse, RE:Survivor with the GunCon. The main issue here is the mouse can go outside the game window (if you play in a window) and is not wrapped inside, what i think is called _Absolute mode_. For the Playstation Mouse there is a _Relative Mode_ option what confines the mouse inside the windows and it looks it works (except for an issue i will explain later) but it only works in the QT version, in the NoGUI there is the _Relative Mode_ option but there is'nt any effect.
The implement of the Relative Mode was reported here https://github.com/stenzek/duckstation/issues/739

**SOLUTION**:

There should be a _Relative Mode_ option for both Playstation Mouse and GunCon controllers, and working for both NoGUI and QT frontends, because the _Relative Mode_ is more about the real mouse and how it is converted to emulate those specific controllers.

**ALTERNATIVES**:

I didn't found any workaround. Only to hide desktop icons and any window to avoid click anything by mistake if the mouse goes outside the window.

**ADDITIONAL CONTEXT**:
* I see an issue with this in the QT version using Relative Mode with Playstation Mouse. The Relative Mode seems to work perfect in the game (C&C:RA), but normally in other apps with the mouse captured in the window, if you want to bring back the mouse to control something outside you press the WINDOWS key (In Windows). In other apps like Virtual Machine Managers, normally they have a keys combination. However if you do that in the mentioned QT version with Playstation Mouse, **the mouse appears in the center stuck, and you can't move the mouse anywhere unless you stop the emulation**. The better solution will be if the windows has lost the focus, return to absolute mode to free the mouse until you focus again the emulator window.
* The Relative Mode could be the default for both Playstation Mouse and GunCon, making an option to disable for absolute mode. The reason is if you play in a window with the Playstation Mouse or GunCon you mostly didn't wanna to click outside by mistake, something very easy to do, but also if you play in Fullscreen because if you have more than one monitor, the mouse will go to the other monitors, meanwhile with one monitor there is no difference with relative and absolute mode.

Thanks!

**Comment by stenzek** [comment]
Not really interested in implementing relative mouse mode in nogui. It's meant for TVs, where there's no windows anyway.

At some point I'm going to add the ability to launch the fullscreen UI in the Qt frontend, so it'll work fine then.

Guncon is absolute positioned input, relative mode doesn't make sense here.

**Comment by LukasThyWalls** [comment]
> Guncon is absolute positioned input, relative mode doesn't make sense here.

I can understand the rest, but this, with the GunCon you can go outside the window at any time? Playing RE: Survivor, being one that could be played with the GunCon but any of the others should be the same, is very frustrating because you can easily click outside the window many times because the mouse is going all over the place in the background. I also clicked things in other monitors playing it.
I mentioned the relative mode because how the Playstation Mouse works in C&C:Red Alert in the QT Version is like how the GunCon controlled games should be working with the Mouse.

**Comment by stenzek** [comment]
To me it makes sense. If it was a real lightgun, you could point it anywhere outside of the TV's display area, just like a mouse pointer.

**Comment by LukasThyWalls** [comment]
> To me it makes sense. If it was a real lightgun, you could point it anywhere outside of the TV's display area, just like a mouse pointer.

I get that, and it make sense to emulate the behaviour of clicking outside the game because is how the GunCon should work, but there should be a difference between the emulated controller and the real controller, here the emulated GunCon and the Real Mouse. If you are playing with the GunCon, you are not freezing/pausing the game everytime you shoot outside and maybe clicking anything in your desktop, but that's what happen now with the actual configuration. With the real GunCon you reload, go forward, nothing or whatever the game was programmed, and nothing else (although for quality of life is good to have a button to do the outside shoot).
Maybe can still be in absolute mode, using the screen full size, but what i'm talking is not allowing somehow the mouse doing things outside the game when the game is playing. That's why i'm talking about wraping the mouse to the game.

> Using fullscreen would be one way to "lock" it to the bounds

But if you are using more than one monitor, in one you have the game in fullscreen but in others the mouse goes and can do things outside the game, no?

**Comment by stenzek** [comment]
You can't "lock the mouse to a window". Relative mode is implemented by constantly moving the mouse to the center of the window every time it moves. This is how PC games do it as well.

**Comment by LukasThyWalls** [comment]
> You can't "lock the mouse to a window". Relative mode is implemented by constantly moving the mouse to the center of the window every time it moves. This is how PC games do it as well.

I know, but also most or almost all the games avoid, somehow, to release the mouse outside the game to not have the issues mentioned above.

I understand you want to simulate better the behaviour of the GunCon shooting outside the emulated screen for the game, but not with the issues of having the real mouse going everywhere, lossing focus and allowing clicking anything in your desktop or in other windows.

**Comment by stenzek** [comment]
I don't wish to discuss this any further. If you feel like implementing it yourself, I'm happy to review/merge, but it's not something I will be doing myself. Doing so would require significant changes to how I handle mouse movement for guncon, and that's not something I have the time nor interest to do.

**Comment by LukasThyWalls** [comment]
Sorry, i didn't wanna to be annoying, pretty far of my intention. I just wanted to bring these topics after testing and doing some things with the emulator, which looks the only what it is being worked on actually and seems better in some respects.

My skills in programming are very basic and i don't have the experience to do this kind of things, so i can't help with that. I have some knowledge of how some things works, and i have some experience with testing.

Anyway i will understand if you want to leave this topic because you don't wanna talk about this anymore, or, if you want, it can be opened a new feature request about this in a more generic point of view (what are the issues and some ideas), summarized and kept it opened. What you want, because is your project.

But i want to put some things i found before more testing:

* I tested this with a multi-monitor setup, and with that doesn't matter if you put the game in fullscreen (on NoGUI or QT), there are issues with the Playstation Mouse with Absolute Mode and the GunCon. The Mouse can go to the other monitors (others than the one is showing the game in fullscreen) and if you click with the mouse in the other monitors, the game exits the fullscreen and goes to window. Then everything works like windowed but the game boundaries of the game cursor (the GunCon crosshair or the Playstation Mouse pointer) goes different to the game viewport, so the ingame cursor doesn't reach the whole game window, like both sizes doesn't match. The only way to fix that is doing ALT+ENTER to reach fullscreen again.
With relative mode with Playstation Mouse and fullscreen, the mouse never leave the monitor with the game, so there is no issue (i didn't test what happens if i go to window with ALT+ENTER, if the mouse goes weird, but it can be tested).

* I found the behaviour of the mouse capture with the relative mode in my two test computers are different with the same DuckStation version (i did the test with the QT version, deattached game window). I'm talking about if i have the game in a window, and i do an alt-tab or press the Windows key, when the game recovers the mouse control. In my main computer (Win10, GTX750, 1 Monitor) the mouse didn't want to leave the game window, unless i alt-tab another window above the game window and hide entirely its content. Then i can move the mouse over my monitor, but if it touch the game viewport, being that window in focus or not, the mouse is captured again. In my other computer (Win10, GTX1660 Super, 2 Monitor), i can ALT-TAB or use the Windows key easily, because the mouse is only captured again if the game window has the focus, so meanwhile it hasn't i can move the mouse everywhere. Why the difference? I don't know.

* Thinking about how the relative mode can work, is there a way to make the mouse leave the relative mode when the game window has lost the focus and when it catch the focus again, put the relative mode? Or maybe put the absolute mode automatically if the window lose the focus, but only go to relative mode again when you click inside the game window.

* What you said of the GunCon should have the ability to shoot outside the game, and idea could be using the mouse in relative mode with the GunCon and add a gap around outside the game what the crosshair can move and can be considered a shoot outside the game, but that gap is not visible in the game viewport. I don't know if this can be done, but with one of the issues mentioned, seems the boundaries of the Playstation mouse in relative mode are somehow calculated to stop in a point, maybe can be done to go a bit far of this and create something with this.

Anyway, thanks for everything.
